src/featureManager.ts

```diff
@@ -30,15 +34,69 @@ export class FeatureManager {
+        return result.variant;
+    }
+
+    async #assignVariant(featureFlag: FeatureFlag, context: ITargetingContext): Promise<{
+        variant: VariantDefinition | undefined;
+        reason: VariantAssignmentReason;
+    }> {
+        // user allocation
+        if (featureFlag.allocation?.user !== undefined) {
+            for (const userAllocation of featureFlag.allocation.user) {
+                if (isTargetedUser(context.userId, userAllocation.users)) {
+                    const variant = featureFlag.variants?.find(v => v.name == userAllocation.variant);
+                    if (variant !== undefined) {
+                        return { variant, reason: VariantAssignmentReason.User };
+                    } else {
+                        console.warn(`Variant ${userAllocation.variant} not found for feature ${featureFlag.id}.`);
+                        return { variant: undefined, reason: VariantAssignmentReason.None };
+                    }
+                }
+            }
         }
 
-        // Ensure that the feature flag is in the correct format. Feature providers should validate the feature flags, but we do it here as a safeguard.
-        validateFeatureFlagFormat(featureFlag);
+        // group allocation
+        if (featureFlag.allocation?.group !== undefined) {
+            for (const groupAllocation of featureFlag.allocation.group) {
+                if (isTargetedGroup(context.groups, groupAllocation.groups)) {
+                    const variant = featureFlag.variants?.find(v => v.name == groupAllocation.variant);
+                    if (variant !== undefined) {
+                        return { variant, reason: VariantAssignmentReason.Group };
+                    } else {
+                        console.warn(`Variant ${groupAllocation.variant} not found for feature ${featureFlag.id}.`);
+                        return { variant: undefined, reason: VariantAssignmentReason.None };
+                    }
+                }
+            }
+        }
 
+        // percentile allocation
+        if (featureFlag.allocation?.percentile !== undefined) {
+            for (const percentileAllocation of featureFlag.allocation.percentile) {
+                const hint = featureFlag.allocation.seed ?? `allocation\n${featureFlag.id}`;
+                if (isTargetedPercentile(context.userId, hint, percentileAllocation.from, percentileAllocation.to)) {
+                    const variant = featureFlag.variants?.find(v => v.name == percentileAllocation.variant);
+                    if (variant !== undefined) {
+                        return { variant, reason: VariantAssignmentReason.Percentile };
+                    } else {
+                        console.warn(`Variant ${percentileAllocation.variant} not found for feature ${featureFlag.id}.`);
+                        return { variant: undefined, reason: VariantAssignmentReason.None };
+                    }

```

Comment:
This could be a reusable function
```suggestion
                    return GetVariantResult(featureFlag, percentileAllocation.variant, VariantAssignmentReason.Percentile);
```

```typescript
GetVariantResult(featureFlag, variantName, VariantAssignmentReason reason);
  const variant = featureFlag.variants?.find(v => v.name == variantName);
  if (variant !== undefined) {
      return { variant, reason: reason };
  } else {
      console.warn(`Variant ${variantName} not found for feature ${featureFlag.id}.`);
      return { variant: undefined, reason: VariantAssignmentReason.None };
  }
}
```

Reply:
this looks good to me, also reducing the tabs/nested blocks.

Reply:
extracted in 63cdd43